### ISSUE TITLE
你这份代码和SLBR的源码有什么区别，模型结构变了吗

### ISSUE THREAD

**Issue by Xelawk** [lianchengmingjue/DENet#2]
能说说修改了哪些地方吗

**Comment by lianchengmingjue** [comment]
模型部分修改主要在文件[src/models/SLBR.py](https://github.com/lianchengmingjue/DENet/blob/main/src/models/SLBR.py)和[src/networks/resunet.py](https://github.com/lianchengmingjue/DENet/blob/main/src/networks/resunet.py)中

**Comment by Xelawk** [comment]
> 模型部分修改主要在文件[src/models/SLBR.py](https://github.com/lianchengmingjue/DENet/blob/main/src/models/SLBR.py)和[src/networks/resunet.py](https://github.com/lianchengmingjue/DENet/blob/main/src/networks/resunet.py)中

你的SharedBottleNeck实现是不是有问题，论文中自注意力是放在下采样与上采样中间，但你的实现是放在了上采样之后，

**Comment by lianchengmingjue** [comment]
这一部分是延续SLBR中share-bottle的操作，即背景修复解码器和水印定位解码器共享第一个上采样层。同样，本仓库中语义特征经过self-attention模块后之后仍然会继续进行后续上采样操作。
由于share-bottle这一部分实现细节比较trivial，在原SLBR论文中也没有将其作为重要创新点，因此本论文中没有体现，而是直接公开代码供感兴趣的读者参考。

**Comment by Xelawk** [comment]
> 这一部分是延续SLBR中share-bottle的操作，即背景修复解码器和水印定位解码器共享第一个上采样层。同样，本仓库中语义特征经过self-attention模块后之后仍然会继续进行后续上采样操作。 由于share-bottle这一部分实现细节比较trivial，在原SLBR论文中也没有将其作为重要创新点，因此本论文中没有体现，而是直接公开代码供感兴趣的读者参考。

原论文的实现没问题，我的意思是你这里的self attention是放在down 和up之后，但你论文中提到的是放在down和up之间。而且你实现的self attention好像也存在问题，正常来说scale是取qkv向量维度的平方根倒数，你这里是取c**(-0.5)，正常来说应该是c // heads的平方根倒数

**Comment by lianchengmingjue** [comment]
因为论文这里如果涉及到share-bottle的介绍会很琐碎。
attention的确当时没有注意到取head头维度做scale这个问题。

**Comment by Xelawk** [comment]
> 因为论文这里如果涉及到share-bottle的介绍会很琐碎。 attention的确当时没有注意到取head头维度做scale这个问题。

你好，方便给个联系方式吗，可以交流下哈哈

**Comment by Xelawk** [comment]
> 因为论文这里如果涉及到share-bottle的介绍会很琐碎。 attention的确当时没有注意到取head头维度做scale这个问题。
还有个问题，我注意到训练流里对水印图片处理有：
```python
        end = time.time()
        bar = Bar('Processing {} '.format(self.args.nets), max=len(self.train_loader))
        for i, batches in enumerate(self.train_loader):
            current_index = len(self.train_loader) * epoch + i

            inputs = batches['image'].float().to(self.device)
            target = batches['target'].float().to(self.device)
            mask = batches['mask'].float().to(self.device)
            
            wm = batches['wm'].float().to(self.device)
            wm = (1-mask)*target +mask*wm  
```
请问为什么要额外做(1-mask)*target +mask*wm  ，而不是直接用wm去提取水印的参考特征，我看论文示例图里也没有这样的操作

**Comment by lianchengmingjue** [comment]
这个步骤在论文第四页中有提到。
In practice, we do not directly feed the watermark image into the share-weight encoder. Instead, we fill the black region of the watermark with the same content as the watermark-free image, otherwise the useless information of 0 will cause damage to the encoder. And here we use Eqn. (3) to remove the content that was filled before.

**Comment by Xelawk** [comment]
> 这个步骤在论文第四页中有提到。 In practice, we do not directly feed the watermark image into the share-weight encoder. Instead, we fill the black region of the watermark with the same content as the watermark-free image, otherwise the useless information of 0 will cause damage to the encoder. And here we use Eqn. (3) to remove the content that was filled before.

感谢解答
